### ISSUE TITLE
Pre-heat feature not working in master

### ISSUE THREAD

**Issue by jinengandhi-intel** [oscarlab/graphene#2172]
## Description of the problem
Preheat enclave feature in the master is not showing any performance improvements and is a regression compared to the first patch (https://github.com/mkow/graphene/commit/089375d450ce5198406ffe6cc234c975c88223af) where we saw massive performance improvements.

Replacing the code in the master with the patch code, again shows the same performance improvements. 

## Steps to reproduce

Run the 3DRESUNET 1 instance test with GSGX with preheat enabled, disabled and replacing the code from patch:

average_time | std | max | min | epoch | compute type | threads | instances
-- | -- | -- | -- | -- | -- | -- | --
4.47 | 3.44 | 14.77 | 3.23 | 10 | preheat_enabled_manifest (latest master) | 20 | 1
4.43 | 3.44 | 14.74 | 3.23 | 10 | preheat_disabled_manifest (latest master) | 20 | 1
3.4 | 0.45 | 4.75 | 3.24 | 10 | preheat enabled by applying the patch | 20 | 1

## Expected results

Preheat feature should show performance improvements on apps like 3Dresunet

**Comment by dimakuv** [comment]
Something is strange here. Both variants of the patch do the same thing:
- https://github.com/mkow/graphene/commit/089375d450ce5198406ffe6cc234c975c88223af -- performs `memset()` over `[heap_min, heap_max)` enclave memory region
- https://github.com/oscarlab/graphene/commit/a402a2a8d9b7649c1fc8a1c9ced47bd5e5111aba -- performs `READ_ONCE()` over `[heap_min, heap_max)` enclave memory region

So these are basically the same... We touch each page in the enclave memory region in both options. @jinengandhi-intel Are you sure the `sgx.preheat_enclave = 1` option was set in the manifest?

**Comment by dimakuv** [comment]
After an offline communication with @jinengandhi-intel , there is indeed a difference between these two variants of the patch.

This is the diff between the two variants:
```
if (preheat_enclave == 1) {
-        for (uint8_t* i = g_pal_sec.heap_min; i < (uint8_t*)g_pal_sec.heap_max; i += g_page_size)
-            READ_ONCE(*i);
+       memset(g_pal_sec.heap_min, 0, g_pal_sec.heap_max - g_pal_sec.heap_min);
}
```

The only difference that I see here is a Read Access (`READ_ONCE`, what is currently in the master) and a Write Access (`memset(0)`, in the other branch). Could this be the culprit? But how?

**Comment by monavij** [comment]
Does not make sense, In both cases the page should be brought in by the driver. So we should not see any difference. Only thing I can think is that driver knows that the page is not mapped and just ignores it? So changing that READ_ONCE to WRITE_ONCE should fix it. But we do need to chat with driver team on this.

**Comment by mkow** [comment]
Fascinating... This doesn't make any sense to me...

As you can see in https://github.com/oscarlab/graphene/pull/2154, I did compare `memset()` and `READ_ONCE()` approaches, and the latter did strictly better (and actually triggered page faults).

@jinengandhi-intel: Did you use exactly mkow@089375d or applied the same patch to the current master? If the former, then you're testing quite a different codebase, there was a bunch of changes in the meantime.

Also, could you try the following experiments?

* Print something to the log in the `READ_ONCE` codepath (to see if the manifest option was really parsed correctly).
* Replace `READ_ONCE` with `WRITE_ONCE`.

**Comment by jinengandhi-intel** [comment]
Did you use exactly mkow/graphene@089375d or applied the same patch to the current master?
>>> We have tried multiple things here, 

1. For the initial set of 3DRESUNET performance numbers we had applied the memset line change to the Graphene master as of Jan26 (back then, this was latest of Graphene master) and the pre-heat changes clearly had a performance improvement.
2. Few days back we wanted to re-collect the performance numbers with the latest master and the last commit here is: [Pal/Linux] Change `ADDR_IN_PAL` to `ADDR_IN_PAL_OR_VDSO` and that's where we noticed that the pre-heat changes don't actually work. 

**So basically the READ_ONCE code in master was not showing any perf improvement, next I replaced the READ_ONCE for loop with the memset line and things started to work.** 

if (preheat_enclave == 1) {
-        for (uint8_t* i = g_pal_sec.heap_min; i < (uint8_t*)g_pal_sec.heap_max; i += g_page_size)
-            READ_ONCE(*i);
+       memset(g_pal_sec.heap_min, 0, g_pal_sec.heap_max - g_pal_sec.heap_min);
}

**Comment by jinengandhi-intel** [comment]
> Replace READ_ONCE with WRITE_ONCE.

```
-        for (uint8_t* i = g_pal_sec.heap_min; i < (uint8_t*)g_pal_sec.heap_max; i += g_page_size)
-            READ_ONCE(*i);
+        log_printf("I am inside preheat enclave code\n");
+       for (uint8_t* i = g_pal_sec.heap_min; i < (uint8_t*)g_pal_sec.heap_max; i += g_page_size)
+            WRITE_ONCE(*i, 0);
```

The above code fixes the issue and I see the preheat feature working where the time taken by epoch 0 is much less and comparable to epoch 1,2 etc.

**Comment by mkow** [comment]
I have no idea how's this possible. Maybe we have a bug in `READ_ONCE()` and the optimizer removed the call to it? Could you disasm this place (in that version which _doesn't_ work) and see how this loop was compiled?

**Comment by mkow** [comment]
Ok, I got the binary from @jinengandhi-intel and it is a bug in our `READ_ONCE()` or GCC (but I think it's the former). Small repro: https://godbolt.org/z/n3WTbz.
I'll analyze this issue further and post more information soon.

**Comment by mkow** [comment]
It seems that we copied out buggy code from Linux:

https://github.com/torvalds/linux/blob/c6a771d932332568df9f46a3b53507c578e8c8e8/include/linux/compiler.h#L182-L194

I tested a 1-1 copy of their READ_ONCE and it's actually broken (see https://godbolt.org/z/hzhP8j), but they `if`-ed out some basic types, so the bug was never detected by them. Basically, the `default` case of that switch just doesn't work.

Interestingly, this bug was accidentally fixed recently when they refactored and simplified the code: https://github.com/torvalds/linux/commit/a5460b5e5fb82656807840d40d3deaecad094044.

I'll submit a PR with a similar simplification, turns out that all this magic they used previously was only to circumvent old, buggy GCC.
